app/src/main/java/com/github/sdp/ratemyepfl/database/reviewable/ReviewableRepositoryImpl.kt

```diff
@@ -0,0 +1,118 @@
+package com.github.sdp.ratemyepfl.database.reviewable
+
+import com.github.sdp.ratemyepfl.database.LoaderRepository
+import com.github.sdp.ratemyepfl.database.LoaderRepositoryImpl
+import com.github.sdp.ratemyepfl.database.Repository
+import com.github.sdp.ratemyepfl.database.RepositoryImpl
+import com.github.sdp.ratemyepfl.database.SearchableRepository.Companion.LIMIT_QUERY_SEARCH
+import com.github.sdp.ratemyepfl.database.query.QueryResult
+import com.github.sdp.ratemyepfl.model.items.Reviewable
+import com.github.sdp.ratemyepfl.model.review.ReviewRating
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.Query
+
+
+/**
+ * Decorator for Repository that defines reviewable related operations
+ *
+ * @param repository: a [LoaderRepositoryImpl] to decorate
+ * @param idFieldName: the name of the field that holds the id of the [Reviewable]
+ */
+class ReviewableRepositoryImpl<T : Reviewable> private constructor(
+    private val repository: LoaderRepository<T>,
+    private val idFieldName: String,
+) : ReviewableRepository<T>, LoaderRepository<T> by repository {
+
+    constructor(repository: Repository<T>, idFieldName: String) : this(
+        LoaderRepositoryImpl(repository),
+        idFieldName
+    )
+
+    constructor(
+        database: FirebaseFirestore,
+        collectionPath: String,
+        idFieldName: String,
+        transform: (DocumentSnapshot) -> T?
+    ) : this(
+        LoaderRepositoryImpl(RepositoryImpl(database, collectionPath, transform)),
+        idFieldName
+    )
+
+    companion object {
+        const val NUM_REVIEWS_FIELD_NAME = "numReviews"
+        const val AVERAGE_GRADE_FIELD_NAME = "averageGrade"
+
+        /**
+         * Compute the updated rating of the provided item if we add the provided rating.
+         *
+         *  @param item: the item for which we compute
+         *  @param rating: rating to add the the item
+         *
+         *  @return a [Pair] containing the computed number of reviews and average grade.
+         */
+        inline fun <reified T : Reviewable> computeUpdatedRating(
+            item: T,
+            rating: ReviewRating
+        ): Pair<Int, Double> {
+            val numReviews = item.numReviews
+            val averageGrade = item.averageGrade
+            val newNumReviews = numReviews + 1
+            val newAverageGrade: Double =
+                averageGrade + (rating.toValue() - averageGrade) / newNumReviews
+
+            return Pair(newNumReviews, newAverageGrade)
+        }
+    }
+
+
+    private val queryMostRated = query()
```

Comment:
Seems like you are not using this.

Reply:
Yes, I wrote it thinking I would have time to implement the filter. I think I won't need it anyway, so I can remove it.